typeclass-experiments/src/Common/SourceTy.hs

```diff
@@ -0,0 +1,166 @@
+{- This is my take on what our proto *should* look like.
+
+Everything that operates on this can easily be adapted to operate on something similar, but
+it will make many functions partial or introduce the need for additional error handling if
+we end up abandoning the use of NonEmpty/etc.
+
+-}
+{-# LANGUAGE LambdaCase #-}
+{- This is my take on what our proto *should* look like.
+
+Everything that operates on this can easily be adapted to operate on something similar, but
+it will make many functions partial or introduce the need for additional error handling if
+we end up abandoning the use of NonEmpty/etc.
+
+-}
+{-# LANGUAGE OverloadedStrings #-}
+{- This is my take on what our proto *should* look like.
+
+Everything that operates on this can easily be adapted to operate on something similar, but
+it will make many functions partial or introduce the need for additional error handling if
+we end up abandoning the use of NonEmpty/etc.
+
+-}
+{-# LANGUAGE PatternSynonyms #-}
+
+module Common.SourceTy where
+
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.List.NonEmpty qualified as NE
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Text (Text)
+
+type NEMap k v = NonEmpty (k, v)
+
+type VarName = Text
+
+type TyConName = Text
+
+newtype TKind = TKind [VarName]
+
+data TypeRef = Local TyConName | Foreign ModuleName TyConName
+
+data Prim0 = TInt | TBool | TString
+
+data Prim1 = TMaybe | TList
+
+data Prim2 = TMap | TEither
+
+data TyPrim
+  = TP0 Prim0
+  | TP1 Prim1
+  | TP2 Prim2
+
+data SourceTy = TyVar VarName | TApp SourceTy SourceTy | TyRef TypeRef | PrimT TyPrim
+
+data TyDef = TyDef
+  { tyDefName :: TyConName
+  , tyDefKind :: TKind
+  , tyDefBody :: TyBody
+  }
```

Comment:
I'm curious why introduce the generator machinery and how does it differ to a vanilla approach by using just-a-function. I see you machinery as a 'rule' system that 'matches' on certain patterns and performs a text conversion, but I don't necessarily see what kind of modularity it brings compared to the following toy approach:

The only 'extension' point is the `opaques` input to `runCodegen` functions where you can add additional opaques as you extend the capabilities.

```haskell
-- These are all running under MonadFail m and MonadState m
-- Imagine the state containing PursModuleName, PursImports and PursExports for the current LB module being codegened

data PursCodegenState = S {
  opaques :: Map TyName (PursModuleName, PursTyName),
  imports :: Map TyName (PursModuleName, PursTyName),
  exports :: Map TyName [Constructor]
}
runCodegen :: Map TyName (PursModuleName, PursTyName) -> CompilerInput -> m Text
runCodegen opaques inp = evalState (S opaques [] [])

toPursModule (Module modName tyDefs _) = do
  setModName modName
  traverse toPursTypeDef tyDefs

toPursTypeDef (TyDef n k b) = do
    n' <- toPursTyName n
    k' <- toPursKind k
    b' <- toPursBody b
    addTyDefExport $ TyDef n k b -- module ModuleName (Foo(..)) etc.
    return $ Text.format "data %s %s = %s" (n', k', b')

toPursTy (TyVar v) = toPurs v
toPursTy (TyApp tyFun tyArgs) = do
    tyFun' <- toPursTy tyFun
    tyArgs' <- toPursTy tyArgs
    return $ Text.format "%s %s" (tyFun', tyArgs')

toPursTy (TyRef r) = do -- Here's where we need an extension option
  ty <- findTyByRef r
  if isOpaque ty then do
    importOpaque ty # this is just consulting a Map TyName (PursModuleName, PursTyName) or raises an error if not found, then adds it to the state PursImports
    
toPursBody (Sum ...) = ...
toPursBody Opaque = ...
```

Reply:
This is a toy as the actual tricky business comes with generating instances and functions, which will likely change this in some ways. But for type def codegen, this seems pretty sufficient.

NOTE: This is a brain dump, I didn't compile it or anything, just a scrabble